llvm/lib/CAS/CachingOnDiskFileSystem.cpp

```diff
@@ -289,34 +364,86 @@ CachingOnDiskFileSystemImpl::makeFile(DirectoryEntry &Parent,
 }
 
 Expected<FileSystemCache::DirectoryEntry *>
-CachingOnDiskFileSystemImpl::lookupOnDiskFrom(DirectoryEntry &Parent,
-                                              StringRef Name) {
+CachingOnDiskFileSystemImpl::lookupOnDiskFrom(
+    DirectoryEntry &Parent, const std::string &TreePath,
+    Optional<sys::fs::file_status> KnownStatus) {
```

Comment:
It feels like there should be an assertion:
```
assert(Parent.getTreePath() == dirname(TreePath));
```
If not, return an error if they mismatch?

Reply:
(I.e., FileSystemCache has an invariant (not sure if it's enforced?) where the tree path of every DirectoryEntry is the tree path of its parent + "/" + its own name. If we're sure we're respecting that, assertion; if not, error and/or recovery.)

Reply:
Good idea. I think we should handle it in `FileSystemCache`, not here. I'll do it separately.

Reply:
Looks like it's already asserted in `makeLazyEntry`.